src/db/row.rs

```diff
@@ -1,58 +1,136 @@
+    NullRemote(RemoteId, NullRecordRow<'conn>),
+    /// The `Records` and `NullRecords` rows do not exist.
+    UndefinedRemote(RemoteId, Undefined<'conn>),
+    /// The alias is undefined.
+    UndefinedAlias(Alias),
+    /// The remote id is invalid.
+    InvalidRemoteId(RecordError),
 }
 
-impl<'conn> RecordsTableRow<'conn> {
-    /// Initialize a new [`RecordsTableRow`] given a transaction and the [`CitationKey`]
-    /// corresponding to the row which either exists or is missing from the `Records` table.
+impl<'conn> State<'conn> {
+    /// Determine the current state of the database, as corresponds to the provided record
+    /// identifier.
     #[inline]
-    pub(super) fn from_tx<K: CitationKey>(
-        tx: Transaction<'conn>,
-        key: &K,
-    ) -> Result<Self, rusqlite::Error> {
-        match get_row_id(&tx, key)? {
+    pub fn determine(tx: Transaction<'conn>, record_id: RecordId) -> Result<Self, rusqlite::Error> {
+        match get_row_id(&tx, &record_id)? {
             Some(row_id) => {
                 debug!("Beginning new transaction for row '{row_id}'.");
-                Ok(RecordsTableRow::Exists(RecordRow::new(tx, row_id)))
-            }
-            None => {
-                debug!("Beginning new generic transaction.");
-                Ok(RecordsTableRow::Missing(MissingRecordRow::new(tx)))
+                Ok(State::Exists(record_id, RecordRow::new(tx, row_id)))
             }
+            None => match record_id.resolve() {
+                Ok(Either::Left(alias)) => {
+                    tx.commit()?;
+                    Ok(State::UndefinedAlias(alias))
+                }
+                Ok(Either::Right(remote_id)) => {
+                    debug!("Checking null row for '{remote_id}'");
+                    match get_null_row_id(&tx, &remote_id)? {
+                        Some(row_id) => {
+                            Ok(State::NullRemote(remote_id, NullRecordRow::new(tx, row_id)))
+                        }
+                        None => Ok(State::UndefinedRemote(remote_id, Undefined::new(tx))),
+                    }
+                }
+                Err(record_error) => {
+                    tx.commit()?;
+                    Ok(State::InvalidRemoteId(record_error))
+                }
+            },
         }
     }
 }
 
-/// A private trait which allows the caller to prepare (cached) sql statements.
-trait DatabaseTransaction {
-    fn prepare(&self, sql: &str) -> Result<rusqlite::Statement, rusqlite::Error>;
+/// A representation of the database state beginning with an arbitrary [`RemoteId`].
+#[derive(Debug)]
+pub enum RemoteState<'conn> {
+    /// The `Records` row exists.
+    Exists(RemoteId, RecordRow<'conn>),
+    /// The `Records` row does not exist and the `NullRecords` row exists.
+    NullRemote(RemoteId, NullRecordRow<'conn>),
+    /// The `Records` and `NullRecords` rows do not exist.
+    UndefinedRemote(RemoteId, Undefined<'conn>),
+    /// The remote id is invalid.
```

Comment:
Maybe it makes more sense to not pass the `RemoteId` back here (and instead just get a reference up-front)? But I wrote it in this way to be consistent with the `State` implementation, where ownership is necessary for the `resolve` step.

Reply:
Not passing it back might make it clearer that the id won't see any change. I feel it's OK either way.

Reply:
I've changed it to not pass the ID back, I think it is a bit more understandable this way.